src/kv/store.h

```diff
@@ -220,80 +220,112 @@ namespace kv
-      {
-        auto search = maps.find(s->get_name());
-        if (search == maps.end())
-        {
-          throw ccf::ccf_logic_error(
-            fmt::format("Map does not exist - {}", s->get_name()));
-        }
-
-        search->second->apply(s);
-      }
-
-      for (auto& map : maps)
+      if (v < commit_version())
       {
-        map.second->unlock();
+        throw ccf::ccf_logic_error(fmt::format(
+          "Attempting to snapshot at version {} which is earlier than "
+          "committed version {} ",
+          v,
+          commit_version()));
       }
 
+      if (v > current_version())
       {
-        std::lock_guard<SpinLock> vguard(version_lock);
-        version = snapshot->get_version();
-        last_replicated = snapshot->get_version();
-        last_committable = snapshot->get_version();
+        throw ccf ::ccf_logic_error(fmt::format(
+          "Attempting to snapshot at version {} which is later "
+          "than current version {}",
+          v,
+          current_version()));
       }
-    }
 
-    std::unique_ptr<AbstractSnapshot> snapshot(Version v) override
-    {
-      std::unique_ptr<AbstractSnapshot> snapshot =
-        std::make_unique<StoreSnapshot>(v);
+      StoreSnapshot snapshot;
 
       {
         std::lock_guard<SpinLock> mguard(maps_lock);
 
-        if (v < commit_version())
-        {
-          throw ccf::ccf_logic_error(fmt::format(
-            "Attempting to snapshot at invalid version v:{}, "
-            "commit_version:{}",
-            v,
-            commit_version()));
-        }
-
         for (auto& map : maps)
         {
           map.second->lock();
         }
 
         for (auto& map : maps)
         {
-          snapshot->add_map_snapshot(std::move(map.second->snapshot(v)));
+          snapshot.add_map_snapshot(map.second->snapshot(v));
         }
 
         for (auto& map : maps)
         {
           map.second->unlock();
         }
       }
-      snapshot->serialize();
-      return snapshot;
+
+      auto e = get_encryptor();
+      KvStoreSerialiser serialiser(e, v, true);
+
+      return snapshot.serialise(serialiser);
+    }
+
+    DeserialiseSuccess deserialise_snapshot(
+      const std::vector<uint8_t>& data) override
+    {
+      auto e = get_encryptor();
+      auto d = KvStoreDeserialiser(e);
+
+      auto v_ = d.init(data.data(), data.size());
+      if (!v_.has_value())
+      {
+        LOG_FAIL_FMT("Initialisation of deserialise object failed");
+        return DeserialiseSuccess::FAILED;
+      }
+      auto v = v_.value();
+
+      std::lock_guard<SpinLock> mguard(maps_lock);
+
+      for (auto& map : maps)
+      {
+        map.second->lock();
+      }
+
+      for (auto r = d.start_map(); r.has_value(); r = d.start_map())
+      {
+        const auto map_name = r.value();
+
+        auto search = maps.find(map_name);
+        if (search == maps.end())
+        {
+          LOG_FAIL_FMT("Failed to deserialize snapshot");
```

Comment:
I think including the version information in the fail message would be useful and doesn't leak anything useful. The map name on the other hand is rightly debug only.